File: library/src/main/java/com/mindera/skeletoid/threadpools/ThreadPoolUtils.java
package com.mindera.skeletoid.threadpools;

import android.support.annotation.VisibleForTesting;

import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * An utils to create thread pools with readable names
 */
public final class ThreadPoolUtils {

    public static final AtomicInteger mThreadTotal = new AtomicInteger(0);

    @VisibleForTesting
    ThreadPoolUtils() {
        throw new UnsupportedOperationException();
    }

    /**
     * Get ThreadPool with fixed number of threads
     *
     * @param threadPoolName The name of the thread pool
     * @param MAX_THREAD     Max number of threads
     * @return Thread Pool
     */
    public static java.util.concurrent.ThreadPoolExecutor getFixedThreadPool(final String threadPoolName, final int MAX_THREAD) {
        return new ThreadPoolExecutor(MAX_THREAD, MAX_THREAD, 0L,
                TimeUnit.MILLISECONDS, new NamedThreadFactory(threadPoolName,
                MAX_THREAD));
    }

    /**
     * Get ScheduledThreadPool with fixed number of threads
     *
     * @param threadPoolName The name of the thread pool
     * @param MAX_THREAD     Max number of threads
     * @return Thread Pool
     */
    public static ScheduledThreadPoolExecutor getScheduledThreadPool(final String threadPoolName, final int MAX_THREAD) {
        return new ScheduledThreadPoolExecutor(MAX_THREAD,
                new NamedThreadFactory(threadPoolName, MAX_THREAD));
    }

    /**
     * Shutdown thread pool and remove number of threads from total count
     *
     * @param threadPoolExecutor The threadpool to be shutdown
     */
    public static void shutdown(java.util.concurrent.ThreadPoolExecutor threadPoolExecutor) {
        if (threadPoolExecutor == null) {
            return;
        }
        mThreadTotal.addAndGet(-threadPoolExecutor.getCorePoolSize());
        threadPoolExecutor.shutdown();
    }

    /**
     * Shutdown thread pool now and remove number of threads from total count
     *
     * @param threadPoolExecutor The threadpool to be shutdown now
     */
    public static void shutdownNow(java.util.concurrent.ThreadPoolExecutor threadPoolExecutor) {
        if (threadPoolExecutor == null) {
            return;
        }
        mThreadTotal.addAndGet(-threadPoolExecutor.getCorePoolSize());
        threadPoolExecutor.shutdownNow();
    }
}

File: base/src/main/java/com/mindera/skeletoid/threads/threadpools/ThreadPoolUtils.kt
package com.mindera.skeletoid.threads.threadpools

import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

/**
 * An utils to create thread pools with readable names
 */
object ThreadPoolUtils {
    @JvmField
    val threadTotal = AtomicInteger(0)

    /**
     * Get ThreadPool with fixed number of threads
     *
     * @param threadPoolName The name of the thread pool
     * @param maxThreads     Max number of threads
     * @return Thread Pool
     */
    @JvmStatic
    fun getFixedThreadPool(
        threadPoolName: String, maxThreads: Int
    ): java.util.concurrent.ThreadPoolExecutor {
        return ThreadPoolExecutor(
            maxThreads,
            maxThreads,
            0L,
            TimeUnit.MILLISECONDS,
            NamedThreadFactory(threadPoolName, maxThreads)
        )
    }

    /**
     * Get ScheduledThreadPool with fixed number of threads
     *
     * @param threadPoolName The name of the thread pool
     * @param maxThreads     Max number of threads
     * @return Thread Pool
     */
    @JvmStatic
    fun getScheduledThreadPool(
        threadPoolName: String,
        maxThreads: Int
    ): ScheduledThreadPoolExecutor {
        return ScheduledThreadPoolExecutor(
            maxThreads,
            NamedThreadFactory(threadPoolName, maxThreads)
        )
    }

    /**
     * Shutdown thread pool and remove number of threads from total count
     *
     * @param threadPoolExecutor The threadpool to be shutdown
     */
    @JvmStatic
    fun shutdown(threadPoolExecutor: java.util.concurrent.ThreadPoolExecutor) {
        threadTotal.addAndGet(-threadPoolExecutor.corePoolSize)
        threadPoolExecutor.shutdown()
    }

    /**
     * Shutdown thread pool now and remove number of threads from total count
     *
     * @param threadPoolExecutor The threadpool to be shutdown now
     */
    @JvmStatic
    fun shutdownNow(threadPoolExecutor: java.util.concurrent.ThreadPoolExecutor) {
        threadTotal.addAndGet(-threadPoolExecutor.corePoolSize)
        threadPoolExecutor.shutdownNow()
    }
}